WD-Firmware/comrv/comrv.c

```diff
@@ -18,34 +18,645 @@
+/* set the comrv stack pool and comrv stack registers */
+#if __riscv_xlen == 64
+ #define M_COMRV_SET_POOL_ADDR(address)  asm volatile ("la t4, "#address : : : ); \
+                                         asm volatile ("ld t4, 0x0(t4)"  : : : );
+ #define M_COMRV_SET_STACK_ADDR(address) asm volatile ("la t3, "#address : : : ); \
+                                         asm volatile ("ld t3, 0x0(t3)"  : : : );
+#elif __riscv_xlen == 32
+ #define M_COMRV_SET_POOL_ADDR(address)  asm volatile ("la t4, "#address : : : ); \
+                                         asm volatile ("lw t4, 0x0(t4)"  : : : );
+ #define M_COMRV_SET_STACK_ADDR(address) asm volatile ("la t3, "#address : : : ); \
+                                         asm volatile ("lw t3, 0x0(t3)"  : : : );
+#endif
+
+/* overlay group size in D_COMRV_OVL_GROUP_SIZE_MIN granularity */
+#define M_COMRV_GET_OVL_GROUP_SIZE(token)          (pOverlayOffsetTable[token.fields.overlayGroupID+1] - \
+                                                    pOverlayOffsetTable[token.fields.overlayGroupID])
+/* overlay group size in bytes */
+#define M_COMRV_GET_OVL_GROUP_SIZE_IN_BYTES(token) ((pOverlayOffsetTable[token.fields.overlayGroupID+1] - \
+                                                     pOverlayOffsetTable[token.fields.overlayGroupID]) << 9)
+/* token offset in bytes */
+#define M_COMRV_GET_TOKEN_OFFSET_IN_BYTES(token)   ((token.fields.offset) * D_COMRV_OFFSET_SCALE_VALUE)
+/* overlay group offset in bytes */
+#define M_COMRV_GET_GROUP_OFFSET_IN_BYTES(token)   ((pOverlayOffsetTable[token.fields.overlayGroupID]) << 9)
 
 /**
 * types
 */
+typedef struct comrvStackFrame
+{
+  /* holds return address (caller is non overlay function) or return
+     offset (caller is overlay function) */
+  u32_t callerReturnAddress;
+  /* holds callee address (callee is non overlay function) or callee
+     token (callee is an overlay function) */
+  u32_t calleeToken;
+  /* holds the offset in bytes to the previous stack frame */
+  s16_t offsetPrevFrame;
+#ifdef D_COMRV_MULTI_GROUP_SUPPORT
+  /* if the calleetoken is a multi group token, this field holds the
+     actual loaded token entry in the multi group table */
+  u16_t calleeMultiGroupTableEntry;
+#endif /* d_comrv_multi_group_support */
+} comrvStackFrame_t;
+
+/* overlay token fields */
+typedef struct comrvToken
+{
+  /* overlay token indication 0: address; 1: overlay token */
+  u32_t overlayAddressToken:1;
+  /* overlay group ID the function resides in */
+  u32_t overlayGroupID:16;
+  /* data/function offset within the overlay group – 4 bytes granularity */
+  u32_t offset:10;
+  /* caller thunk indication */
+  u32_t thunk:1;
+  /* reserved */
+  u32_t reserved:1;
+  /* specify the heap ID this overlay group belongs to */
+  u32_t heapID:2;
+  /* multi group indication */
+  u32_t multiGroup:1;
+} comrvTokenFields_t;
+
+/* overlay token */
+typedef union comrvOverlayToken
+{
+  u32_t              value;
```

Comment:
add TODO comment: check without bitfield

Reply:
done in 4a8e389